docs/middleware.md

```diff
@@ -683,6 +683,39 @@ to use the `middleware=<List of Middleware instances>` style, as it will:
 * Ensure that everything remains wrapped in a single outermost `ServerErrorMiddleware`.
 * Preserves the top-level `app` instance.
 
+## Applying middleware to `Mount`s
+
+Middleware can also be added to `Mount`, which allows you to apply middleware to a single route, a group of routes or any mounted ASGI application:
+
+```python
+from starlette.applications import Starlette
+from starlette.middleware.gzip import GzipMiddleware
+
+
+routes = [
+    Mount(
+        "/",
+        routes=[
+            Route(
+                "/example",
+                endpoint=...,
+            )
+        ],
+        middleware=[GzipMiddleware]
+    )
+]
+
+app = Starlette(routes=routes)
+```
+
+Note that middleware used in this way is *not* wrapped in exception handling middleware like the middleware applied to the `Starlette` application is.
+This is often not a problem because it only applies to middleware that inspect or modify the `Response`, and even then you probably don't want to apply this logic to error responses.
+If you do want to apply the middelware logic to error responses only on some routes you have a couple of options:
+
+* Add an `ExceptionMiddleware` onto the `Mount`
+* Add a `try/except` block to your middleware and return an error response from there
+* Split up marking and processing into two middlewares, one that gets put on `Mount` which simply marks the response as needing processing (for example by setting `scope["log-response"] = True`) and another applied to the `Starlette` application that does the heavy lifting.
```

Comment:
Hmm... This works... 🤔 

```python
from starlette.applications import Starlette
from starlette.middleware import Middleware
from starlette.responses import Response
from starlette.routing import Mount, Route


async def home(request):
    return Response("Hi there!")


async def exception_handler(request, exc):
    return Response("I have a 400 for you!", status_code=400)


class PotatoException(Exception):
    ...


class CustomMiddleware:
    def __init__(self, app):
        self.app = app

    async def __call__(self, scope, receive, send):
        raise PotatoException()


routes = [
    Mount(
        "/",
        routes=[
            Route(
                "/example",
                endpoint=home,
            )
        ],
        middleware=[Middleware(CustomMiddleware)],
    ),
]
app = Starlette(routes=routes, exception_handlers={PotatoException: exception_handler})
```

Reply:
Try this example:

```python
from starlette.applications import Starlette
from starlette.middleware import Middleware
from starlette.responses import Response
from starlette.routing import Mount, Route

class PotatoException(Exception):
    ...

async def bad_endpoint(request):
    raise PotatoException

async def exception_handler(request, exc):
    return Response("I have a 400 for you!", status_code=400)

class CustomMiddleware:
    def __init__(self, app, name: str):
        self.app = app
        self.name = name

    async def __call__(self, scope, receive, send):
        async def wrapped_send(msg):
            if msg["type"] = "http.response":
                print(f"{self.name} called for {scope['raw_path']}")
            await send(msg)
        await self.app(scope, receive, wrapped_send)

routes = [
    Mount(
        "/mount",
        routes=[
            Route(
                "/bad",
                endpoint=bad_endpoint,
            )
        ],
        middleware=[Middleware(CustomMiddleware, name="on_mount")],
    ),
    Route(
        "/good",
        endpoint=bad_endpoint,
    )
]

app = Starlette(
    routes=routes,
    exception_handlers={PotatoException: exception_handler},
    middleware=[Middleware(CustomMiddleware, name="on_app")],
)
```

The "on_mount" middleware will never be called because PotatoException tears through `Mount`'s middleware stack. On the other hand "on_app" always gets called because it's "protected" by ErrorMiddleware.

Reply:
Ah, got it. But that's not what I got from what's written here...

Do you think an image with middleware/app/mount as "blocks" would be helpful understanding this? 🤔

(I can help if you think it makes sense, but I lack design skills to do it 😎 👍)

Reply:
That sounds like a good idea, please give it a shot! I won't have time until next week (and also don't have the design skills 😆). Maybe do it with ascii text instead of an image since it will be easier to embed?

Reply:
This is not a blocker. 👍
I'll not have time to do this, if someone eventually wants, feel free to open a PR with it. 👍 

(Please do not resolve this conversation so ppl see it)

Reply:
I'm not sure I understand why requests to sub-mounted routes bypass middleware associated with that mount? Isn't `ErrorMiddleware` supposed to wrap the _entire_ `Starlette.router`, which includes sub-apps?

Reply:
I got this when I ran the snippet above: https://github.com/encode/starlette/pull/1649/files#r960241635

Reply:
Lemme try to draw this out to make sure I understand...

Looking at the Starlette code, we have this layering:

```python
routes = [
    Route("/", ...),
    Mount("/sub", sub, middleware=[Middleware(MountMiddlewareA)]),
]

app = Starlette(
    routes=routes,
    middleware=[Middleware(MiddlewareB), Middleware(MiddlewareC)],
)
```

```python
ServerErrorMiddleware(
  MiddlewareC(
    MiddlewareB(
      ExceptionMiddleware(
        Router(
          routes=[
            Route("/", ...),
            Mount("/sub", MiddlewareA(sub))
          ]
        )
      )
    )
  )
)
```

So, if an exception occurs in `/`, it gets handled by `ExceptionMiddleware` and `Middleware{B,C}` get a well-defined response _[perhaps an error response]_ to process.

But if an exception occurs in `/sub*`, it gets sent to `MiddlewareA`, which has to handle it if it hopes to do some response processing (otherwise that processing isn't run).

Is that right?

Reply:
Yes, that's right.

Adrian's example was missing something (I had to add it myself before, and I forgot)...

```python
from starlette.applications import Starlette
from starlette.middleware import Middleware
from starlette.responses import Response
from starlette.routing import Mount, Route

class PotatoException(Exception):
    ...

async def bad_endpoint(request):
    raise PotatoException

async def exception_handler(request, exc):
    return Response("I have a 400 for you!", status_code=400)

class CustomMiddleware:
    def __init__(self, app, name: str):
        self.app = app
        self.name = name

    async def __call__(self, scope, receive, send):
        async def wrapped_send(msg):
            if msg["type"] == "http.response.start":  # IT WAS MISSING `.start`
                print(f"{self.name} called for {scope['raw_path']}")
            await send(msg)
        await self.app(scope, receive, wrapped_send)

routes = [
    Mount(
        "/mount",
        routes=[
            Route(
                "/bad",
                endpoint=bad_endpoint,
            )
        ],
        middleware=[Middleware(CustomMiddleware, name="on_mount")],
    ),
    Route(
        "/good",
        endpoint=bad_endpoint,
    )
]

app = Starlette(
    routes=routes,
    exception_handlers={PotatoException: exception_handler},
    middleware=[Middleware(CustomMiddleware, name="on_app")],
)
```

Reply:
We can use @florimondmanca 's drawing here for the explanation 👍

Reply:
Sorry for the broken example 😫, but yes you got it right Florimond!

Reply:
I added a test in aec580f

Reply:
@florimondmanca I'm realizing now that it was not clear if this was completely resolved, sorry if I missed it. Do you think we should tweak the docs more, maybe adding something along the lines of your explanation in https://github.com/encode/starlette/pull/1649#discussion_r962144793?